python/ffsim/protocols/trace_protocol.py

```diff
@@ -45,3 +51,72 @@ def trace(obj: Any, norb: int, nelec: tuple[int, int]) -> float:
         "The object did not have a _trace_ method that returned the trace "
         "or a _diag_ method that returned its diagonal entries."
     )
+
+
+def _trace_term(
+    op: tuple[tuple[bool, bool, int], ...], norb: int, nelec: tuple[int, int]
+):
+    n_elec_alpha, n_elec_beta = nelec
+
+    combined_indices = set([(orb, spin) for _, spin, orb in op])
+    norb_alpha = sum([not spin for _, spin in combined_indices])
+    norb_beta = sum([spin for _, spin in combined_indices])
+
+    eta_alpha = 0
+    eta_beta = 0
+
+    # loop over the support of the operator
+    # assume that each site is either 0 or 1 at the beginning
+    # track the state of the site through the application of the operator
+    # if the state exceed 1 or goes below 0,
+    # the state is not physical and the trace must be 0
+    for i, spin in combined_indices:
+        initial_zero = 0
+        initial_one = 1
+        is_zero = True
+        is_one = True
+        for action, aspin, orb in reversed(op):
+            if (aspin, orb) != (spin, i):
+                continue
+
+            change = action * 2 - 1
+            initial_zero += change
+            initial_one += change
+            if initial_zero < 0 or initial_zero > 1:
+                is_zero = False
+            if initial_one < 0 or initial_one > 1:
+                is_one = False
+            if not is_zero and not is_one:  # no possible initial state
+                return 0j
+        # if neither the state 0 nor 1 returns to the initial state, the trace must be 0
+        if (is_zero and initial_zero != 0) or (is_one and initial_one != 1):
+            return 0j
+        # only one case is possible if the operator has support on site i
+        assert not is_zero or not is_one
+        # count the number of electrons
+        if is_one:
+            if not spin:
+                eta_alpha += 1
+            else:
+                eta_beta += 1
+        # the number of electrons exceed the number of allowed electrons
```

Comment:
```suggestion
        # the number of electrons exceeds the number of allowed electrons
```
Also, I think this comment belongs _inside_ the following if block.